### DISCUSSION TITLE
Can't display more than 2 swatches for quantitative data

### DISCUSSION THREAD

**Discussion by stevage** [observablehq/plot#827] [Q&A] [answered]
I'm not sure if this is a bug, a feature request or support request.

I have a dataset which has approx 10 discreet values, integers from 0 to 15.

I'd a like a legend to display a reasonable number of those values (most likely, all of them). I'm setting it up like this:

```js
const plotEl = Plot.plot({
  marks: [
    Plot.rectY(
        expeditions,
        Plot.binX(
            { y: 'count' },
            {
                x: {
                    value: 'date',
                    interval: d3.utcMonth,
                },
                fill: 'participantsCount'
            }
        )
    ),
  ],
  background: '#222',
  width: document.getElementById('chart').getClientRects()[0].width,
      style: {
          background: '#222',
          color: 'white',
      },
  });
  const legendEl = plotEl.legend('color', {
      legend: 'swatches',
      columns: 8,
      style: { color: 'white', background: '#222' },
  });
  document.getElementById('chart').replaceChildren(plotEl);
```

I can't see anything in the docs that offers a way to display more swatches.

Result:

![image](https://user-images.githubusercontent.com/678344/160018729-ab2f56a3-c3ce-484f-9f10-032da3fb1ca6.png)

Of note:

- only two swatches (min and max value)
- also some strange artefacts that look like a 1 pixel height swatch

See in action [here](http://mapswell.site44.com/geohashing-swatches/) (select Show chart, Number of participants).

**Comment by mbostock** [comment, marked as answer]
The _swatches_ legend type is only supported for discrete quantitative color scales, but you’re using a continuous quantitative color scale here. (We should probably throw an error in this case.) Per the README:

> For quantitative color scales, it defaults to ramp but may be set to swatches for a discrete scale (most commonly for threshold color scales); for ordinal color scales and symbol scales, only the swatches value is supported.

The _swatches_ legend type will display all the values in the scale’s domain, which in the case of a continuous scale is the lower and upper bound, which appears to be [0, 15] in your case.

The default legend for a linear color scale is:

<img width="309" alt="Screen Shot 2022-03-24 at 3 43 37 PM" src="https://user-images.githubusercontent.com/230541/160022094-2f6b291b-55b7-4a7c-a3d7-93e86f8b20a3.png">

```js
Plot.legend({
  color: {
    type: "linear"
  }
})
```

But with _swatches_ you just see 0 and 1:

<img width="286" alt="Screen Shot 2022-03-24 at 3 43 04 PM" src="https://user-images.githubusercontent.com/230541/160022038-a7f09162-7fc8-49ce-a87a-ee59e732c077.png">

```js
Plot.legend({
  legend: "swatches",
  color: {
    type: "linear"
  }
})
```

If you want to show a discrete color legend, you need to switch to a discrete color scale so that the legend matches the scale. For example, you could use a _threshold_ scale:

<img width="282" alt="Screen Shot 2022-03-24 at 3 45 28 PM" src="https://user-images.githubusercontent.com/230541/160022302-edb10d43-7b3b-4e23-bff0-f996348e7bb6.png">

```js
Plot.legend({
  color: {
    type: "threshold",
    domain: d3.range(1, 14)
  }
})
```

A threshold scale will default to the _rdylbu_ color scheme but you can switch to _turbo_ if you prefer that.

<img width="292" alt="Screen Shot 2022-03-24 at 3 45 50 PM" src="https://user-images.githubusercontent.com/230541/160022336-923e38ff-cd86-4caf-8a8e-d3ab0785595c.png">

```js
Plot.legend({
  color: {
    type: "threshold",
    scheme: "turbo",
    domain: d3.range(1, 14)
  }
})
```

You could also use a _quantile_ scale, but this changes the semantics of the color encoding.

If Plot supported _quantize_ scales then you wouldn’t need to compute the thresholds yourself. I filed [#828](https://github.com/observablehq/plot/issues/828) for that.

**Reply by stevage** [reply]
Thanks, that was a really helpful explanation. Yes, I think `threshold` is the one I want.

**Reply by stevage** [reply]
Although, this doesn't seem to actually work:

```js
            const legendEl = plotEl.legend('color', {
                legend: 'threshold',
                domain: d3.range(1, 14),
                style: { color: 'white', background: 'transparent' },
            });
```

("Unknown legend type: 'threshold'")

Setting a `type` property instead of `legend` had no effect. Passing an object with `{ type: 'legend' }` was a type error.

**Reply by stevage** [reply]
And doing it this way did work:

```js
const legendEl = Plot.legend({
                color: {
                    type: 'threshold',
                    domain: d3.range(1, 14),
                },
                style: { color: 'white', background: 'transparent' },
            });
```

I'm not sure why the difference in form is significant.

**Reply by stevage** [reply]
Ok, I half-get it. 

In the second case, `color: { }` is actually defining a whole new scale, and `type` is the type of that scale, and the legend is just happening automatically.

In the first case, 'color' is identifying an existing scale, and `legend` is the type of the *legend*, not the scale.

**Reply by mbostock** [reply]
Yep, you got it!
